src/contractors.jl

```diff
@@ -0,0 +1,125 @@
+# contractors:
+"""
+    Contractor{F}
+
+    Abstract type for contractors.
+"""
+abstract type Contractor{F} end
+
+export Bisection, Newton
+
+# bisection
+struct Bisection{F} <: Contractor{F}
+    f::F
+end
+
+function (contractor::Bisection)(X, tol)
+    image = contractor.f(X)
+
+    if !(contains_zero(image))
+        return :empty, X
+    end
+
+    return :unknown, X
+end
+
+# Newton
+struct Newton{F,FP,O} <: Contractor{F}
+    f::F
+    f_prime::FP
+    op::O
+end
+
+function Newton(f::Function, f_prime::Function)
+    Newton(f, f_prime, 𝒩)
+end
+
+function (C::Newton)(X, tol)
+    # use Bisection contractor for this:
+    if !(contains_zero(IntervalBox(C.f(X))))
+        return :empty, X
+    end
+
+    # given that have the Jacobian, can also do mean value form
+
+    NX = C.op(C.f, C.f_prime, X) ∩ X
+
+    isempty(NX) && return :empty, X
+
+    if NX ⪽ X  # isinterior; know there's a unique root inside
+        NX =  refine(X -> C.op(C.f, C.f_prime, X), NX, tol)
+        return :unique, NX
+    end
+
+    return :unknown, NX
+end
+
+
+# What is this guarded_mid for? Shouldn't it be checking if f(m)==0?
```

Comment:
This comment is confusing (I am picky I know, but the PR is about cleaning up ^^)

Reply:
I think we should just get rid of guarded_mid.